crates/fuzzing/src/oracles/memory.rs

```diff
@@ -0,0 +1,316 @@
+        if let Some(n) = len
+            .checked_sub(usize::try_from(offset).unwrap())
+            .and_then(|n|
+            // Only convert to a `u32` after the `checked_sub` above, because
+            // `len` can validly be `2**32` while `u32` can only hold up to
+            // `2**32 - 1`.
+            u32::try_from(n).ok())
+        {
+            for i in 0..=1 {
+                let addr = n + i;
+                assert!(
+                    load8.call(&mut *store, addr).is_err(),
+                    "{msg}: len={len:#x}, offset={offset:#x}, load8({n} + {i:#x} = {addr:#x}) traps"
+                );
+                assert!(
+                    load16.call(&mut *store, addr).is_err(),
+                    "{msg}: len={len:#x}, offset={offset:#x}, load16({n} + {i:#x} = {addr:#x}) traps"
+                );
+                assert!(
+                    load32.call(&mut *store, addr).is_err(),
+                    "{msg}: len={len:#x}, offset={offset:#x}, load32({n} + {i:#x} = {addr:#x}) traps"
+                );
+                assert!(
+                    load64.call(&mut *store, addr).is_err(),
+                    "{msg}: len={len:#x}, offset={offset:#x}, load64({n} + {i:#x} = {addr:#x}) traps"
+                );
+            }
+        }
+
+        // Test out-of-bounds accesses near the end of the index type's range to
+        // double check our overflow handling inside the bounds checks.
+        let len_is_4gib = u64::try_from(len).unwrap() == u64::from(u32::MAX) + 1;
+        let end_delta = len_is_4gib as u32;
+        let max = u32::MAX;
+        for i in 0..(1 - end_delta) {
+            let addr = max - i;
+            assert!(
+                load8.call(&mut *store, addr).is_err(),
+                "{msg}: len={len:#x}, offset={offset:#x}, load8({max:#x} - {i:#x} = {addr:#x}) traps"
+            );
+        }
+        for i in 0..(2 - end_delta) {
+            let addr = max - i;
+            assert!(
+                load16.call(&mut *store, addr).is_err(),
+                "{msg}: len={len:#x}, offset={offset:#x}, load16({max:#x} - {i:#x} = {addr:#x}) traps"
+            );
+        }
+        for i in 0..(4 - end_delta) {
+            let addr = max - i;
+            assert!(
+                load32.call(&mut *store, addr).is_err(),
+                "{msg}: len={len:#x}, offset={offset:#x}, load32({max:#x} - {i:#x} = {addr:#x}) traps"
+            );
+        }
+        for i in 0..(8 - end_delta) {
+            let addr = max - i;
+            assert!(
+                load64.call(&mut *store, addr).is_err(),
+                "{msg}: len={len:#x}, offset={offset:#x}, load64({max:#x} - {i:#x} = {addr:#x}) traps"
+            );
+        }
+    };
+
+    do_accesses(&mut store, "initial size");
+    let _ = memory.grow(&mut store, u64::from(growth));
+    do_accesses(&mut store, "after growing");
+}
+
+fn build_wasm(image: &HeapImage, offset: u32) -> Vec<u8> {
```

Comment:
Do you think it's worth doing this in the text format instead of expanding out all the wasm-encoder bits?

Reply:
Ha, I originally had that and rewrote it into the text format to get a little bit better throughput for the fuzzer. I can add a comment giving an overview of the generated module or something to make readability a bit better at a glance though.